### ISSUE TITLE
Optimizing choices given

### ISSUE THREAD

**Issue by The-furf-of-July** [antialiasis/favorite-picker#21]
I am writing this under the assumption that the options to select from are chosen at random. If this is incorrect, take this suggestion with a grain of salt.

I think a method to improve the amount of time a user needs to interact with the system to get a result could be improved if the choices provided are at least slightly weighted to be "meaningful." If the pool of pokémon to choose from is very large, this is especially important.

A bit of explanation to show what I mean by meaningful choices:
- In a given group of pokémon, there are likely a handful of strongly preferred pokémon. These are the desired result. There are also a large number of pokémon that the user is indifferent towards or dislikes.
- By searching through the collected data so far, you can check how many others have been eliminated by a given pokémon. If a pokémon has been the victor in several different decisions, it can be implied the user has a decent opinion of them. If there are no or few results for a given pokémon beating others, then we don't have a lot of information on that pokémon.
- If we list an entry we have a lot of information about with entries that we do have information about, there are two outcomes. Either the low-information entry gets eliminated, or we get more information about it by it eliminating something else. 

Additionally, the user will have an easier time choosing between something they like (something that we know keeps winning) and something they are indifferent towards, compared to say, nothing but strongly preferred choices, or nothing but indifferent choices. 

Near the end of a round, all of the entries in the pool are preferred by the user to some degree. However, we can still compare the ones that have won more to the ones that have won the least, which is far more likely to provide the user with a straightforward decision and accurate results than fully random selection.
Due to the nature of those end-of-round choices, a lot of their top choices are going to get pitted against each other early on, which hides them until the victor gets added to the favorites list. If we give them choices they are less likely to be as strongly conflicted about, then the ones they feel the strongest about remain in the pool till the very end.

- It is a self-sustaining cycle. By having a preference for avoiding pitting favorites against each other until the end, we'll have more entries with more data in the pool. The more high-info entries we have, the easier it is to randomly select one they like, and compare it to entries we don't know much about.
- This also helps prevent indifferent entries from staying in the pool till the end simply due to random chance, since the selection is making it a priority to learn more about the entries that haven't been shown.
- In large pools, information is collected much faster, and entries that haven't been shown yet are slightly more likely to be compared to things we already know a bit about. This improves the odds of either learning more about the old entry by it continuing to win, or more about the new entry by extension (it just beat an entry with 10 wins, so is now 10 entries closer to the top)

As for implementation, like I said near the start I feel like this should be weighted chances. By still having some RNG, the user still gets the option to reroll. I'm not saying to hard code it to always pick the entries with the least and most information, but to have some sort of bias towards including a bit of both ends of the spectrum.

By making it more likely to ask the user straightforward and impactful questions, we improve both accuracy, time efficiency, and ease of use.

**Comment by antialiasis** [comment]
Apologies for the late response, I guess I never got around to responding when you first asked this. Thanks for the thoughtful ticket!

The way the picker determines which Pokémon to show you is by taking the entire list of non-eliminated Pokémon, shuffling it, and then showing you the first n Pokémon from the list, then the next n Pokémon after that, and so on. It is not randomized for each chunk.

This also means that until we've picked a top favorite we always have roughly the same amount of information about every Pokémon. On the first go round, you haven't expressed a preference about any of these Pokémon before. Once you've been through the whole list, we take the Pokémon you picked and reshuffle them; now you're looking at the Pokémon that you have picked exactly once. After that round, you're looking at Pokémon you've picked twice. And so on.

The sole exception is that if at the end of a round we don't have enough Pokémon left for a full batch, we currently shuffle the remaining ones into the list of Pokémon you picked for the next round (I might ultimately alter how this works, though, by e.g. taking the last three hypothetical batches and redistributing them to have roughly the same number of Pokémon). So you will see a *small* percentage of Pokémon here and there that you haven't picked as often as the others. But they will already tend to be present with mostly Pokémon you have already picked, since they will be a small percentage (and the odds that a Pokémon will happen to be one of these leftover few for *multiple* rounds are even smaller). In general, we don't really *have* any extra information during this whole first (and longest by far) run that could tell us which of these Pokémon are more or less likely to be in your top few, so we don't have anything to go on to let us arrange the Pokémon in a way that would support that.

You're probably thinking more of after you've picked a top favorite, when you get to pick between the Pokémon eliminated by that top favorite - some of which may have been ones that were originally eliminated in the first round because they happened to get pitted against the top favorite at the start (this will only happen if you picked *only* the top favorite from that batch, mind, since if you picked the top favorite *and* anything else in that batch the others won't reappear until *both* of them have been added to the favorite list), and some of which will have survived multiple comparisons previously before getting eliminated. But this run after picking the top favorite will be a *whole lot* smaller and less time-consuming anyway! If you start off picking between every Pokémon with no forms (just over a thousand items), then you'll have a likely maximum of about a few dozen Pokémon in this second run, depending on your picking behaviour (I ran some simulations, and if you always pick exactly one favorite from each batch, you'll have about 32 Pokémon in this second run; if you may pick anywhere from one to five the *average* is closer to fifteen but it *can* go up to like 45, if the batches with your top favorite in them from the first couple rounds happen to be some of the ones where you pick exactly one; if you may pick up to ten the average inches further down but the absolute highest maximum I managed to get over more than a thousand simulations was 62).

So I'm not sure that implementing this technique for these post-first-run matchups would ultimately save all that much time compared to the random shuffle; you'll basically always have your second favorite in like max five picks from there anyway!

It is also possible to imagine a different algorithm for the batching of the *first* run, where, say, if you pick some number of Pokémon from the batch, it'll show you those Pokémon again for the next batch but swap out all the other ones. At a glance, this would probably be a somewhat faster way to pick your top few favorites, while giving less information to continue from there - which is a valid use case for the picker, of course, if all you *want* is a top few, though I also suspect it'd kind of encourage you to just keep clicking the same Pokémon without really looking at the other ones, and it sounds intuitively a bit repetitive to me. But it might be fun to try implementing that algorithm, and maybe some would prefer it!

**Comment by The-furf-of-July** [comment]
I may have not communicated my idea very effectively, let me try to clarify.
> In general, we don't really have any extra information during this whole first (and longest by far) run that could tell us which of these Pokémon are more or less likely to be in your top few, so we don't have anything to go on to let us arrange the Pokémon in a way that would support that.

We are able to measure how many other items have been ranked lower than a given item, or something approximately equivalent to a win count.
When I referred to an entry with "a lot of information", I am referring to one that has eliminated more entries than usual. Low-information entries would be ones that have not eliminated many entries at all.

If we track how often a given choice eliminates another entry, then we can see how it is doing and how often the user has favored it so far. Note, this does not (yet) indicate anything _solid_ about its placement on the favorites list, nor is it trying to. It is not whether it is ranked more highly per se, it's just that it has appeared in a larger number of successful decisions, which is a good thing for the user compared to passing.

Having the picker prioritize entries with unusually high or low numbers of eliminations has several good effects, including moving unseen entries to the front sooner, and keeping options the user (seems to/potentially) prefer in play.

For context, a rather common occurence in my usage of this tool, especially in early rounds, has been: there are a handful of options that just immediately win hands down, and dozens of options that I am indecisive or indifferent about, which frequently get passed on to avoid giving inaccurate long-term decisions. I'd only end up making choices if a "heavy hitter" shows up, in which case it tends to eliminate everything else. Since it's all randomised, the odds of one of these showing up is fairly low, and if I ever made a decision that pitted one of these pokémon against each other, the pool of things I'd want to pick would get even smaller and I'd have to reroll even more. As a result, I try saving those choices to the end, once the easy eliminations were done.
This is a specific case, sure, but my proposed idea should apply to more situations than that. Any situation where a user frequently hits pass for a given item or set of items, we'd be able to tell because it has a lower than average "win" count. If we give the user a choice between a few things that haven't eliminated anything, and a few that have a lot of eliminations, the choice should be much more likely to be straightforward.

Tracking the wins of each entry would be fairly straightforward, either check how many times it appears in "eliminated by" sections, or keep a running tally whenever an eliminated entry is added or removed that mentions it, adding and subtracting so you know how many entries are below it.

There's also the option of allowing multiple sorting settings, and letting the user choose whether the list is pure RNG, sequential, strategic, etc.

> So I'm not sure that implementing this technique for these post-first-run matchups would ultimately save all that much time compared to the random shuffle; you'll basically always have your second favorite in like max five picks from there anyway!

I suppose it is worth noting that this isn't necessarily to reduce the number of matches that need to be made, though it may help with that too. The primary intentions are to (1) reduce the need for the user to reroll/pass on a batch, and (2) try to help decisions stay more consistent over long sessions by making batches more likely to have easier choices.

A use case for the picker that is currently difficult to do (especially with my decisionmaking issues) is making longer lists. The method holds up very well for the top picks, but as the selection process continues, a lot of misinformation can slip into the dataset unless the user is very picky about their selections. Being picky with selections results in a lot more rerolling, and a lot more waiting on RNG to give something you're confident on. Giving the picker some form of optional strategy (letting heavy hitters take longer to be eliminated, and making sure entries that haven't showed up yet make it to the front sooner rather than later, etc) would help with this.

> It is also possible to imagine a different algorithm for the batching of the first run, where, say, if you pick some number of Pokémon from the batch, it'll show you those Pokémon again for the next batch but swap out all the other ones. At a glance, this would probably be a somewhat faster way to pick your top few favorites, while giving less information to continue from there - which is a valid use case for the picker, of course, if all you want is a top few, though I also suspect it'd kind of encourage you to just keep clicking the same Pokémon without really looking at the other ones, and it sounds intuitively a bit repetitive to me. But it might be fun to try implementing that algorithm, and maybe some would prefer it!

That alternate first round method does sound like a pretty interesting one, Perhaps make it have two confirm buttons, one that keeps the current selections and rerolls the rest, and the other that rerolls all and lets you start again. This would help mitigate tedium or autopilot, as the user can start another batch if their attention span runs low.

**Comment by antialiasis** [comment]
> If we track how often a given choice eliminates another entry, then we
can see how it is doing and how often the user has favored it so far.

I understood your idea; what I was trying to explain is that during the
entire first round (note that a “round” is not the same thing as the entire
run until you find your first favorite), *none of the entries you see have
eliminated any other entry*. When you either pick an entry *or* pick
another entry over it, you won’t see that entry again until the next round,
so every batch you see is a batch of Pokémon that *haven’t* eliminated
anything - we don’t have more information on some than others. You will
always see every Pokémon - or nearly every, with the exception of any last
few that don’t make a full batch - before you see any of them again; one
round through all of the non-eliminated Pokémon is a “round”. As intended,
you pick something from each batch and the next round shows you only the
Pokémon that you picked from the previous round, so the second round is
Pokémon you have picked once, the third is Pokémon you have picked three
times, and so on, until you end up with a single favorite.

Another bit of your message revealed the real source of the disconnect,
though: you’re doing a lot of just passing on most batches. My reasoning
about the second round being much shorter doesn’t apply if you’re passing
on most batches, because passing is equivalent to picking the entire batch,
passing them all on to the next round. So if you pass on everything but a
few batches, obviously the second round will indeed have almost every
Pokémon in it but only some of them have eliminated anything, and in a case
like that it would theoretically make sense to have some kind of logic
about which get matched against which based on that. But this isn’t really
the intended way to use the picker, so I’m a little hesitant about
programming it around accommodating that case. Passing is there as a last
resort and is not meant to be the routine choice, hence why the “pass”
button is very deemphasized next to “pick”; it’s *especially* not meant to
be the routine choice in the first round, where you’re being presented with
twenty whole Pokémon, and I really can’t recommend using it that way. (If
you *do* find it putting a couple of your ‘heavy hitters’ in the same
batch, though - what you really probably want to do is to just pick both of
them and eliminate the rest, rather than pick one of the two! That’s
something you can already control.)

The concern that you have about ‘misinforming’ the picker is of course a
real one - the picker can only work with what you give it, and in practice
humans don’t have a singular, well-defined, consistently ordered list of
preferences, so things are liable to be a bit fuzzy and making a hasty pick
among a bunch of Pokémon doesn’t necessarily mean you genuinely have a
*consistent* preference for the one you picked over the ones you didn’t.
There are limits to how well it can account for that without simply
throwing out the concept entirely, but this is why the picker has the
ability to reorder your picks after they’ve been picked onto the list. I
have also been musing on whether it would make sense to have a “rescue”
feature where you can get back a Pokémon that you mistakenly eliminated at
some point, though it’s difficult to imagine that working by any other
means than you happening to consciously notice you haven’t been seeing a
Pokémon you think ought to be on your list.

Either way, I might go for implementing that alternative batching or
something along those lines for people with your kind of reluctance to pick
among less-preferred Pokémon. Thanks for sharing your thoughts!

**Comment by The-furf-of-July** [comment]
Ah, I didn't notice the usage of "round" to mean that.
Admittedly, even acknowledging that the system wouldn't work for the first round, I still see benefits.

I typically do make a lot of eliminations on the first round, but do keep in mind, with forms enabled, there are well over a thousand entries, and with cautious picking (not passing!) a large number of choices will make it into round 2. And as the rounds go on, especially with cautious and deliberate selections, the difficulty of choices goes up significantly. The proposed system would in theory get increasing value the more indecisive and selective the user is, and make narrowing down the last few easier by trending towards the highest and lowest win counts of the survivors. (ex. "This entry has won the least out of what's left, let's see if it gets eliminated")

I have also noticed an issue where, I suspect that as it transitions to the next round and shuffles again, some pokémon can get sent to the back prematurely when they haven't been shown, or get less representation than they should. And in rounds with that many entries (forms enabled), having to wait 2 rounds for something to show up is very significant.

It would also allow noticing the difference between a group of 8 pokémon that eliminated 1 each, versus the group of 2 pokémon that eliminated 7 pokémon in a single round. Counting wins would still notice if given pokémon tend to win in smaller groups (higher preference) or larger ones (general preference). Admittedly this isn't _as_ valuable as the system can be with passing. 
But again, passing gets more tempting the tougher the choices get, so even what info it does collect may help at the end.

As stated before, it could be an optional setting, or have it tuned so the bias is more or less present in the shuffling, and the win tallying itself should be relatively straightforward if the data structures are tweaked a little. (When eliminating something, add its win count to the things it was eliminated by, when bringing something back, subtract the win count from each)

With the things you mentioned I can definitely see this being a concept with lower priority than I expected, but I still think it is a viable one for several reasons.

> There are limits to how well it can account for that without simply
throwing out the concept entirely, but this is why the picker has the
ability to reorder your picks after they’ve been picked onto the list. 

Part of why I'm so insistent on using the picker, is to have a systematic method to figure out relatively-objectively how much I like each pokémon.
...My (honestly rather unrealistic) long term goal would be to rank _all_ the pokémon. Which is... not exactly very feasible without adjustments to the system. Hell, it gets tricky in the top 20 sometimes.
Hence my emphasis on long term info gathering and keeping the ones that are doing well near the top so they don't push each other too far down.
Because it's not just about getting through the round, it's about packing as much useful and confident information into the save as I can so it _stays_ accurate.

> I have also been musing on whether it would make sense to have a “rescue”
feature where you can get back a Pokémon that you mistakenly eliminated at
some point, though it’s difficult to imagine that working by any other
means than you happening to consciously notice you haven’t been seeing a
Pokémon you think ought to be on your list.

Definitely add this. I have been doing this via save editing, it'd be a welcome addition.

> Either way, I might go for implementing that alternative batching or
something along those lines for people with your kind of reluctance to pick
among less-preferred Pokémon. Thanks for sharing your thoughts!

I am sure any added feature you make will be a welcome one!
I've tried poking at the code myself, but I'm incredibly rusty and haven't had much luck. Plus, I'm missing the dataset and haven't figured out any way to download it.
